Circuit and Cell Events (Final Draft)

teor2345 opened the pull request:
This pull request is here to remind me to merge the code.

teor2345 committed:
Add new CELL and CIRCUIT events, and placeholders for other expected events

Part of #368.

teor2345 committed:
Add a placeholder function for the new CIRCUIT event

Refactor the old CIRCUIT event so it's generated via the new event's
function.

Part of #368.

teor2345 committed:
Add a placeholder function for the new CELL event

Add some missing entry points, we'll filter them out in the PrivCount
code if we need to.

Part of #368.

teor2345 committed:
Remove all instances of relay_command

It's now cell->command.

Part of #368.

teor2345 committed:
Always send cell events when the circuit is NULL

There won't be any circuit fields in these events, so ounters can filter
them out.

Part of #368.

teor2345 committed:
Mark circuits for close before sending the cell event

And explain how that is handled in the event function

teor2345 committed:
When a received relay cell fails decryption, add fields to the event

teor2345 committed:
Document that cell events include origin cells

teor2345 committed:
Don't send duplicate cell events in command_process_relay_cell()

teor2345 committed:
Consistently order circuit actions, channel actions, and cell events

Always mark circuts for close before sending a cell event.

Always send destroy cells on channels after sending the received cell's event

teor2345 committed:
Add extra circuit end events for origin circuits

Continue to send legacy circuit events just as they were sent in previous
versions.

teor2345 committed:
Add validation for control_event_privcount_circuit_cell() arguments

teor2345 committed:
Add fields to the cell event

teor2345 committed:
Consistently use inbound/outbound and cells/bytes in identifiers

teor2345 committed:
Fix line lengths after the search and replace

teor2345 committed:
Initialise try_relay_command

This is safer if more cases are added later.

teor2345 committed:
Add legacy circuit fields to the new circuit close event

Also refactor out common cell and circuit fields into their own
function.

teor2345 committed:
Rename buf to fingerprint in privcount_list_hs_v2_intro_point_fingerprints()

teor2345 committed:
Add the fingerprint and relay flags for previous and next nodes

Don't add nickname and IPv4 address, they are redundant (and can be
retrieved from the consensus if needed).

Don't add anything for this relay: the aggregator already knows this
relay's flags and fingerprint.
(And a lot of other info, check the get_* functions for details.)

teor2345 committed:
Add some missing consts to the HSDir Store event function

teor2345 committed:
Put the legacy exit bandwidth fields at the end of the event

teor2345 committed:
Add state, purpose, and spliced rend circuit IDs to the circuit event

Don't put channel IDs in each cell event.
Rearrange some fields for clarity.

Part of #368.

teor2345 committed:
Make NULL handling better, again

teor2345 committed:
Add cell counts and directory byte counts to the circuit event

Also refactor the calling code so it's much simpler, and updates the
counters for legacy and new events.

teor2345 committed:
Cleanup some whitespace

teor2345 committed:
Remove some unused functions

teor2345 committed:
Redesign the relay position flags for hidden services

Typical circuits are:
Origin -> Entry -> Mid+ -> End
Origin -> Entry&End

Where End is one of:
* Exit (excluding Dir and HSDir, they're internal)
* Dir
  * including directory mirror and bridge directory circuits, but
  * excluding DirPort, because DirPort connections have no circuits
* HSDir (store and fetch, check flags for which one)
* Intro (client and service, check purpose for which one)
* Rend (client and service, check purpose for which one)

As a special case, clients may extend unsuccessful intro circuits to
another intro point, leading to circuits like this:

Origin -> Entry -> Mid+ -> Mid/Intro* -> End/Intro

Or for Tor2web:

Origin -> Entry/Intro -> Mid/Intro* -> End/Intro

Also clarify a comment, and tidy up some whitespace.

robgjansen reviewed (approved):
Glad to see that the positions were clarified, since I think that info will be extremely useful to have on the PrivCount side.

teor2345 committed:
Report HiddenServiceVersionNumber in HSDir and Intro circuit events

We don't know the HiddenServiceVersionNumber on rend points: their
handshake cell is padded (and possibly encrypted).

teor2345 committed:
Add missing prefixes to the Circuit Id fields

teor2345 committed:
Fix some non-standard field names

Also:
* remove some redundant fields
* move the cell EventTimestamp to its standard location
* fix some comments

teor2345 committed:
Fix commenting conventions for proposed events for test_event_match.sh

teor2345 committed:
Stop trying to count directory bytes at the origin
